presto-main/src/main/java/io/prestosql/operator/scalar/DateTimeFunctions.java

```diff
@@ -139,6 +146,28 @@ public static long fromUnixTime(@SqlType(StandardTypes.DOUBLE) double unixTime,
         return packDateTimeWithZone(Math.round(unixTime * 1000), zoneId.toStringUtf8());
     }
 
+    @ScalarFunction("from_unixtime_nanos")
+    @SqlType("timestamp(9)")
+    public static LongTimestamp fromUnixTimeNanos(@SqlType(StandardTypes.BIGINT) long unixTimeNanos)
+    {
+        long epochSeconds = unixTimeNanos / Timestamps.NANOSECONDS_PER_SECOND;
+        long nanosOfSecond = unixTimeNanos % Timestamps.NANOSECONDS_PER_SECOND;
```

Comment:
This needs to use `floorDiv` and `floorMod` to produce correct results when the input is negative. In particular, nanosOfSecond should never be negative

Reply:
ah interesting, do we need to consider negative unix timestamps as valid? I would prefer to throw when they are encountered.

Reply:
I don't see why not. 

From wikipedia (https://en.wikipedia.org/wiki/Unix_time)

> The Unix time number is zero at the Unix epoch, and increases by exactly 86400 per day since the epoch. Thus 2004-09-16T00:00:00Z, 12677 days after the epoch, is represented by the Unix time number 12677 × 86400 = 1095292800. This can be extended backwards from the epoch too, using negative numbers;

Reply:
👍 ok will do